### PR TITLE
Prevent token collection corruption by fixers

### PR DESCRIPTION
**Pull request by stof**
This prevents replacing token content with unrelated content (hacky fix) in places which should override tokens. Given that the Tokens collection is cached and will be reused by the following fixer, a corrupted collection could break the following fixer as it expects token types to be valid.

The goal is to prevent having to deal with cases like that: https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1097#discussion_r27307639

Replaces #1098

### CODE REVIEW

File: Symfony/CS/Fixer/Contrib/HeaderCommentFixer.php [hunk 1]
```diff
@@ -91,7 +91,7 @@ private static function encloseTextInComment($header)
         foreach ($lines as $line) {
             $comment .= rtrim(' * '.$line)."\n";
         }
-        $comment .= " */\n";
+        $comment .= ' */';
```
Comment (GrahamCampbell) [inline Symfony/CS/Fixer/Contrib/HeaderCommentFixer.php]: You need to rebase surely?

**Comment by stof** [comment]
The remaining test failure seems to be related to the fact that WhitespacyCommentTransformer is not part of the 1.7 codebase.

File: Symfony/CS/Fixer/Contrib/HeaderCommentFixer.php (hunk 1 above)
Comment (stof) [inline Symfony/CS/Fixer/Contrib/HeaderCommentFixer.php]: rebase ? why ?

**Comment by stof** [comment]
@keradus we have a warning in the testsuite related to the LintManager here.

**Comment by keradus** [comment]
> @keradus we have a warning in the testsuite related to the LintManager here.

ref #1152

File: Symfony/CS/Fixer/Contrib/HeaderCommentFixer.php (hunk 1 above)
Comment (keradus) [inline Symfony/CS/Fixer/Contrib/HeaderCommentFixer.php]: No. If rebase would be needed then the merge wouldn't be possible.

This change was indeed done, but on master branch.

Comment (GrahamCampbell) [inline Symfony/CS/Fixer/Contrib/HeaderCommentFixer.php]: I was just saying that surely this change was already made on the 1.7 branch, so it doesn't need fixing again?

Comment (keradus) [inline Symfony/CS/Fixer/Contrib/HeaderCommentFixer.php]: As I said..
that change was NOT made on on 1.7 branch: https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/1.7/Symfony/CS/Fixer/Contrib/HeaderCommentFixer.php#L94
it was made on master branch: https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/master/Symfony/CS/Fixer/Contrib/HeaderCommentFixer.php#L97

**Comment by keradus** [comment]
:+1:

Please, get rid of that merge commits, eg by rebasing.
Also, there are multiple authors, could you do the squashing manually?

This will be a problem on our 2.0 line. After this PR will be merged I will merge 1.7 into master, then disable that new tests, and then send a PR for master with enabling the tests again, this way the 2.0 won't be always broken on Travis and we will have a PR to work on fixing token collection corruption on 2.0.

File: Symfony/CS/Fixer/Contrib/HeaderCommentFixer.php (hunk 1 above)
Comment (GrahamCampbell) [inline Symfony/CS/Fixer/Contrib/HeaderCommentFixer.php]: Ok, then. What I was getting at is that this is unrelated, and should already be fixed on 1.7. I assumed it was you see.

**Comment by stof** [comment]
@keradus should we really squash commits refactoring the different fixers ?

**Comment by keradus** [comment]
Why not? If there where all yours then they will be squashed automatically during gh merge command.
After squashing we will have case that is already often - one person send a commit that proves bug and the other one send a commit that fixes it.

**Comment by stof** [comment]
done

**Comment by keradus** [comment]
Do you plan to do sth more in this PR ?
I am asking in WIP flag context.

**Comment by stof** [comment]
@keradus no. Given that tests are passing, this is not WIP anymore. I removed the flag

**Comment by keradus** [comment]
Great. Thank you!
Your PR brings an fantastic idea!

**Comment by keradus** [comment]
Thank you @stof.

**Comment by keradus** [comment]
btw... WOW, 18k of assertion now :D

**Comment by gharlan** [comment]
> btw... WOW, 18k of assertion now :D

nice :D

**Comment by keradus** [comment]
just a little tool to help with CS ;D
